cirq/protocols/json_serialization.py

```diff
@@ -623,23 +622,23 @@ def to_json(
         class ContextualEncoder(cls):  # type: ignore
             """An encoder with a context map for concise serialization."""
 
-            # This map is populated gradually during serialization. An object
-            # with components defined in this map will represent those
+            # These lists populate gradually during serialization. An object
+            # with components defined in 'context' will represent those
             # components using their keys instead of inline definition.
-            context_map: Dict[str, 'SerializableByKey'] = {}
+            context: List['SerializableByKey'] = []
+            keys: List[str] = []
 
             def default(self, o):
-                skey = getattr(o, '_serialization_key_', lambda: None)()
-                if skey in ContextualEncoder.context_map:
-                    if ContextualEncoder.context_map[skey] == o._json_dict_():
-                        return _SerializedKey(o)._json_dict_()
-                    raise ValueError(
-                        'Found different objects with the same serialization key:'
-                        f'\n{ContextualEncoder.context_map[skey]}\n{o}'
-                    )
-                if skey is not None:
-                    ContextualEncoder.context_map[skey] = o._json_dict_()
-                return super().default(o)
+                if not hasattr(o, '_serialization_name_'):
+                    return super().default(o)
+                try:
```

Comment:
we could convert this to a for loop for linear search - that would be a bit more readable / less fragile than a try/except

Reply:
Included in "hacked version".